cpp/src/Ice/Object.cpp

```diff
@@ -203,82 +198,98 @@ Ice::Object::_iceCheckMode(OperationMode expected, OperationMode received)
     }
     else
     {
-        in.writeParamEncaps(&outEncaps[0], static_cast<int32_t>(outEncaps.size()), ok);
+        incoming.writeParamEncaps(&outEncaps[0], static_cast<int32_t>(outEncaps.size()), ok);
     }
     return true;
 }
 
 bool
-Ice::BlobjectArray::_iceDispatch(Incoming& in, const Current& current)
+Ice::BlobjectArray::_iceDispatch(Incoming& incoming)
 {
+    const Current& current = incoming.current();
     pair<const uint8_t*, const uint8_t*> inEncaps;
     int32_t sz;
-    in.readParamEncaps(inEncaps.first, sz);
+    incoming.readParamEncaps(inEncaps.first, sz);
     inEncaps.second = inEncaps.first + sz;
     vector<uint8_t> outEncaps;
     bool ok = ice_invoke(inEncaps, outEncaps, current);
     if(outEncaps.empty())
     {
-        in.writeParamEncaps(0, 0, ok);
+        incoming.writeParamEncaps(0, 0, ok);
     }
     else
     {
-        in.writeParamEncaps(&outEncaps[0], static_cast<int32_t>(outEncaps.size()), ok);
+        incoming.writeParamEncaps(&outEncaps[0], static_cast<int32_t>(outEncaps.size()), ok);
     }
     return true;
 }
 
 bool
-Ice::BlobjectAsync::_iceDispatch(Incoming& in, const Current& current)
+Ice::BlobjectAsync::_iceDispatch(Incoming& incoming)
 {
     const uint8_t* inEncaps;
     int32_t sz;
-    in.readParamEncaps(inEncaps, sz);
-    auto async = IncomingAsync::create(in);
-    ice_invokeAsync(vector<uint8_t>(inEncaps, inEncaps + sz),
-                    [async](bool ok, const vector<uint8_t>& outEncaps)
-                    {
-                        if(outEncaps.empty())
-                        {
-                            async->writeParamEncaps(0, 0, ok);
-                        }
-                        else
+    incoming.readParamEncaps(inEncaps, sz);
+    auto incomingPtr = make_shared<Incoming>(std::move(incoming));
+    try
+    {
+        ice_invokeAsync(vector<uint8_t>(inEncaps, inEncaps + sz),
+                        [incomingPtr](bool ok, const vector<uint8_t>& outEncaps)
                         {
-                            async->writeParamEncaps(&outEncaps[0], static_cast<int32_t>(outEncaps.size()), ok);
-                        }
-                        async->completed();
-                    },
-                    async->exception(), current);
+                            if(outEncaps.empty())
+                            {
+                                incomingPtr->writeParamEncaps(0, 0, ok);
+                            }
+                            else
+                            {
+                                incomingPtr->writeParamEncaps(&outEncaps[0], static_cast<int32_t>(outEncaps.size()), ok);
+                            }
+                            incomingPtr->completed();
+                        },
+                        [incomingPtr](std::exception_ptr ex) { incomingPtr->completed(ex); }, incomingPtr->current());
+    }
+    catch (...)
+    {
+        incomingPtr->failed(std::current_exception());
```

Comment:
Does `ice_invokeAsync` throw any exception synchronously?

Reply:
This is really about following the same pattern for all async dispatches.

And yes, BlobjectAsync::ice_invokeAsync is a pure virtual function and it's possible for its implementation to throw an exception synchronously. Same for BlobjectArrayAsync::ice_invokeAsync.